### DISCUSSION TITLE
Validation error message wrong/poor for root level discriminated union

### DISCUSSION THREAD

**Discussion by peterbplanetly** [fastapi/fastapi#8578] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Annotated, Union, Literal

import pydantic
import fastapi

# fastapi==0.68.1
# pydantic[dotenv]==1.9.0

class Cat(pydantic.BaseModel):
    type: Literal["cat"]


class Dog(pydantic.BaseModel):
    type: Literal["dog"]


Pet = Annotated[Union[Cat, Dog], pydantic.Field(discriminator="type")]

# pydantic.parse_obj_as(Pet, {"type": "frog"})
#
# No match for discriminator 'type' and value 'frog' (allowed values: 'cat', 'dog')
# (type=value_error.discriminated_union.invalid_discriminator; discriminator_key=type; discriminator_value=frog; allowed_values='cat', 'dog')


app = fastapi.FastAPI()


@app.post("/pet", status_code=201)
def create_pet(pet: Pet):
    return


# Request:
# {
#     "type": "frog"
# }
#
# Response:
# {
#     "detail": [
#         {
#             "loc": [
#                 "body",
#                 "type"
#             ],
#             "msg": "unexpected value; permitted: 'cat'",
#             "type": "value_error.const",
#             "ctx": {
#                 "given": "frog",
#                 "permitted": [
#                     "cat"
#                 ]
#             }
#         },
#         {
#             "loc": [
#                 "body",
#                 "type"
#             ],
#             "msg": "unexpected value; permitted: 'dog'",
#             "type": "value_error.const",
#             "ctx": {
#                 "given": "frog",
#                 "permitted": [
#                     "dog"
#                 ]
#             }
#         }
#     ]
# }
```

### Description

See example code. Pydantic lets us define and parse root level discriminated unions via `Annotated[Union[...], pydantic.Field(discriminator="...")]` + `parse_obj_as`. Pydanitc gives a useful error message in the case of a validation error where no match can be found for the discriminator. When calling the FastAPI endpoint the error message is much less informative. 

### Operating System

Linux

### FastAPI Version

0.68.1

### Python Version

3.9.7

**Comment by sshishov** [comment]
What is the outcome from this?

**Comment by YuriiMotov** [comment]
This was fixed.

Now it gives:
```
{
  "detail": [
    {
      "type": "union_tag_invalid",
      "loc": [
        "body"
      ],
      "msg": "Input tag 'frog' found using 'type' does not match any of the expected tags: 'cat', 'dog'",
      "input": {
        "type": "frog"
      },
      "ctx": {
        "discriminator": "'type'",
        "tag": "frog",
        "expected_tags": "'cat', 'dog'"
      }
    }
  ]
}
```
